### ISSUE TITLE
General inquiry

### ISSUE THREAD

**Issue by BastiaanBekooij** [manny405/sapai#1]
I was curious if you are still working on this project? It's something I was thinking about doing too. My main motivator would be to learn what pets are actually good and to find new synergies.

Since I'm a master's student with Python and ML knowledge this project fits closely to my capabilities. Therefore I'm secretly wondering if I could contribute to the project :)

The difficulty is probably on how to stimulate long-term rewards. Do you agree?

**Comment by BastiaanBekooij** [comment]
That's great to hear! I will dive deeper into the code later and see if I can add something!

I like the idea of iteratively going through the tiers. It might however give a bias towards food related builds and high lvl animals. But that makes it even more interesting :)

**Comment by derBobo** [comment]
I have a suggestion for updating the pet dataset,maybe you could use a api.json file generated from bencoveney/super-auto-pets-db

**Comment by manny405** [comment]
@derBobo You are correct. The pet dataset in current use is obtained from this project and also available from  [https://superauto.pet/api.json](https://superauto.pet/api.json). I've copied the file and included the license from the linked project in sapai/data.py. In the future, it will be better to just this file from the web-page to automatically apply game patches.

**Comment by clsandoval** [comment]
I might be wrong here but intuition tells me one constraint with training an RL agent for something like SAP without an initial phase of imitation learning will be computation. Maybe we can get a "pro" player to record some playtime with this environment. 

On another note, emergent synergies that aren't intuitive sound cool. Maybe a tournament style league could bring out novel strategies? Something similar to Deepmind's Alphastar league? Ill look over the code later and definitely to add any value I can. 

Happy to see I wasn't the only one who thought of this after playing SAP. :D

**Comment by manny405** [comment]
@clsandoval Thanks for the input! I'm finishing up a major overhaul of the entire code-base on the new_patch branch. Give me a day or two to push everything to the main branch. From there, it'd be nice to get some more eyes on testing the behavior of different team compositions during battles to make sure that the behavior of the game is faithfully reproduced. The fight-graph feature is a nice way to test this. 

For SAP, I think that training a pure RL agent right-away is unlikely to lead to good results without many powerful servers dedicated to training. 

I think instead a natural starting place is to develop a way to evaluate how favorable a certain team-state where the team-state includes the round, pets, stats, equipment, and order of the team. The team-state will take a value between 0 and 1.
This way, a simple agent can be written that makes actions to maximize the team-state during the shopping phase. There is still a combinatorial issue that evolves from how many actions deep the agent can search.

Building a way to evaluate the team-state starts with building a database of team compositions for every round and battling them against each other. All possible teams can be evaluated up to probably round 3. Then, smarter algorithms for increasing the size of the database to further rounds will have to be implemented. In the end, a neural-network, which includes a pet-embedding layer (similar to word2vec), will be trained using this database to evaluate how favorable a certain team-state is. The advantage of the NN is that removes the reliance on the database by hopefully being transferable to teams that have not previously been seen before. With this NN and pet-embedding methods, training RL models will become more reliable, thus requiring less compute to achieve reasonable AI behavior.

**Comment by sportshead** [comment]
Wow it seems like so many people are coming to SAP and deciding to create an AI agent in order to play the game! I'm one of them, and I actually started work on creating a model in TypeScript using tfjs, until I stumbled onto this repo while looking for an API for SAP 🤣 

My initial thought was to use a DQN based on imitation learning - I think that it doesn't have to be a pro player, since SAP doesn't have much of a leaderboard/ranking system in place. I've been so far trying to figure out how to do OCR combined with image sectioning in order to read the values off of an image. I think that using popular streamer's vods are a good way to start, especially since many of them are available on Youtube. Filtering out those games that result in a loss would be a good cherry picking method, however some valid strategies may be foiled by one singular round of bad rng. However, because of the massive abundance of vods available from different streamers such as Northernlion and Ludwig that are regarded to be "good" players in the community, this should not be that hard of a task.

I have massive respect for what you have done so far @manny405, please forgive me if I am being ignorant as I have never formally studied the field of ML. I'm also very interested in helping with contribution with this agent system, just have to dust off the python skills!

P.S.: I named my agent `SAAP`, the _Super Auto Auto Pets_, which you can consider using instead of `sapai` 😉

**Comment by clsandoval** [comment]
@manny405 

That sounds like a good approach, if the agent will take on the entire search space for the first 3 turns, the policy can be evaluated to develop state values. Am I correct in understanding that the current plan is for the neural network to act as the state value function in this case? If so, the policy can be as simple as taking the legal action to move in direction of better state value.

As mentioned earlier I think the challenge here would be to represent long term rewards. Either way, I'll read up on using networks as function approximators for state values and will add what I can once the new codebase is up.

This seems really promising, if we can get the agent to properly represent and estimate the value of team compositions in the first 3 rounds, it would be possible to go with a more classical search approach like MCTS. Although in that case it would be model based and the probabilities of units appearing would need to be properly represented.

**Comment by clsandoval** [comment]
> 

General use OCR packages will have a hard time reading the values off of the screen for SAP, I had success doing OCR for bloons tower defense 6 with this model [ https://arxiv.org/abs/1806.10447](url). However, I could not find the font for SAP so it was a dead end on that front sadly.

**Comment by sportshead** [comment]
@clsandoval Digging through the game asset files reveals that the game uses Lapsus Pro Bold as the font. It seems to be [licensed for free use.](https://www.1001fonts.com/lapsus-pro-font.html#license)

**Comment by clsandoval** [comment]
@sportshead  Thanks for the info, Ill update you if I can make it work.

**Comment by manny405** [comment]
@sportshead DQN model will work. The problem with a lot of complex models is initial training and generalization. A lot of popular ML applications (AlphaGo/AlphaZero) don't really talk about this. This is because they already have large existing datasets of expert level play and they don't have restricted training resources (GPU/TPU). If a good team-state approximator can be built initially for the DQN model, it should significantly speed up the performance of the model during training. 


@clsandoval MCTS will be great for propagating the database past round 3. Are there good existing Python libraries? That would save a lot of time. One concern is how to stored training data efficiently. It'd be better if others had already implemented that. It'd save a lot of time.

@sportshead @clsandoval I'll get a move on patching everything I'm working on up today. The previous code-base (on the main branch) for fighting/battles turned into absolute spaghetti code. I've made the new API more consistent and effective for applying effects from abilities. The new code will be much easier for others to contribute. I'll also convert my testing scripts to unit testing on github to help us as well for evaluating merge-requests.

**Comment by clsandoval** [comment]
@manny405 The python libraries I am aware of that have MCTS implemented [https://github.com/pbsinclair42/MCTS](url) only handle the search/search depth, not currently aware of any that tackle efficient storage of training data. Ill ask around and post an update if I find anything.

**Comment by manny405** [comment]
@clsandoval A popular repository I saw was this one [https://github.com/suragnair/alpha-zero-general](https://github.com/suragnair/alpha-zero-general). Might also be a good place to start.

**Comment by manny405** [comment]
@clsandoval @sportshead @BastiaanBekooij @derBobo The huge update has been pushed to the main branch! Please see the commit notes for more details. 

If there are any bugs/errors in any of your individual, please either make a new issue post or, preferable, a patch and a pull request. As stated in this post [issues_6](https://github.com/manny405/sapai/issues/6), please use the testing scripts to ensure that no new errors have been introduced by any changes.

**Comment by manny405** [comment]
Let's open new specific threads for continued conversion about any of these topics when new discussion points come-up.
